### ISSUE TITLE
Enabling safari driver on MacOS LUCI bots

### ISSUE THREAD

**Issue by nturgut** [flutter/flutter#57758]
Our integration tests relies on browser drivers to be installed/started (along with the browsers)

For Chrome Linux tests we solved this issue by using CIPD packages. MacOS on the other hand comes with build in Safari and SafariDriver with one caveat. We need to enable safari driver once after reboot:
```
/usr/bin/safaridriver --enable
```
When I wanted to run this step on luci I got the following error ([link](https://chromium-swarm.appspot.com/task?id=4c514fdb93544210))
```
Password:Password is not valid, please try again
```

I also tried to use sudo and got the error:
```
sudo: no tty present and no askpass program specified
```

Can we configure the macOS luci bots to enable safari driver during the configuration?

**Comment by nturgut** [comment]
@godofredoc @digiter Is there an update on this task? I want to add NOPASSWD permission for /usr/bin/safaridriver --enable. Which will enable me to run, Flutter Driver tests and Flutter Web integration tests on LUCI.

I also added a similar comment to: https://github.com/flutter/flutter/issues/58012#issuecomment-638395106

**Comment by godofredoc** [comment]
Sudo is not allowed on bots. Is this a command that can be run once when the machine boots?

**Comment by nturgut** [comment]
Yes we can run "/usr/bin/safaridriver --enable" when the machine boots. It only needs to run once.

**Comment by nturgut** [comment]
@digiter @godofredoc shall assign this one to infra team? This is a config change that bots should run on startup.

**Comment by digiter** [comment]
1. Is this about the try bots or the prod bots? We control the try bots but chrome team controls the prod bots.
2. What user account needs to run this command? The root user?

**Comment by nturgut** [comment]
1. Both of them since we are planning to use this for Safari integration tests (both mobile and desktop safari)
2. The user who runs the tests (I don't know if it's the root user)

**Comment by digiter** [comment]
Filed http://fxb/54298 for the chrome hosted bots.

**Comment by digiter** [comment]
@godofredoc We have already allowed several sudo nopasswd commands, see http://shortn/_6v6AhGEAfs. If the swarming user needs to run this command, it's easier to add it to the sudoers list. Otherwise the fix will be very complex as our salt config currently doesn't support running a command on startup.

**Comment by nturgut** [comment]
@digiter @godofredoc adding ` "/usr/bin/safaridriver --enable"` to the list @digiter seems very reasonable to me. Thanks for the suggestion.

**Comment by godofredoc** [comment]
@digiter sounds good, let's add it to flutter bots only. That will unblock try and we''ll need to wait for an answer from chromium ops for prod.

**Comment by digiter** [comment]
Created http://tqr/221653 to add it to the sudoers list.

**Comment by yjbanov** [comment]
Has the tqr change propagated to the LUCI bots?

**Comment by yjbanov** [comment]
Can we close the issue?

**Comment by nturgut** [comment]
@yjbanov Last time it took half a week. I will close the issue once it's working. Thanks.

**Comment by godofredoc** [comment]
The change is ready on try pool. Can you please give it a try.

**Comment by nturgut** [comment]
I tested the change, it passed in prod bots:  https://chromium-swarm.appspot.com/task?id=4cfc9da804baaa10

It failed in try bots: https://chromium-swarm.appspot.com/task?id=4cfc6024c2fb5810&o=true

I have one example from each though.

/cc @digiter is there a difference between two configs. Looks like the try config still needs /usr/bin/safaridriver --enable to before we start it.

**Comment by nturgut** [comment]
I added @digiter again to help me debug the difference between the two. 

Thanks a lot for your help!

**Comment by godofredoc** [comment]
@digiter do we need to reboot them for the config to take place?

**Comment by digiter** [comment]
My change is for the try bots. I just applied the change to all the try bots. Please try again.
I don't know why the prod bots worked, the tracking bug is http://tqr/221653.

**Comment by godofredoc** [comment]
@digiter can you please share the steps you used to apply the changes?

**Comment by digiter** [comment]
I logged onto salt master and applied `state.highstate`.

**Comment by nturgut** [comment]
Thanks @digiter Safari driver worked fine for me for try bots as well.

**Comment by nturgut** [comment]
Unfortunately I'm reopening the issue @digiter May be the solution is as easy (ex: rebooting the config files)

Looks like not all the bots have the same configuration. For example this bot has the safari driver enabled so I can run it: https://chromium-swarm.appspot.com/task?id=4d72e564c6f4ff10

However this other bot didn't have safaridriver enabled: https://chromium-swarm.appspot.com/task?id=4d87c0f53d327510

They are both in try pool, I don't have access to prod pool with led anymore.

**Comment by digiter** [comment]
Although we have enabled sudo on `/usr/bin/safaridriver --enable`, we still need to actually run it. Have you tried adding it to the web engine recipe?

**Comment by digiter** [comment]
From another thread with @nturgut , we can help you run `/usr/bin/safaridriver --enable` command on "flutter-try-mac-X" bots. Does it sound good to you?

**Comment by nturgut** [comment]
Now I'm trying to import configurations to Safari as a plist. I think it will solve both issues (this one and https://github.com/flutter/flutter/issues/61971)

The workaround worked in my machine. I want to test it in the try bots but it looks like I don't have access to cipd packages right not. Can I have access to flutter-internal cipd packages? As @godofredoc  suggested before I uploaded other files I manage there: Firefox, Chrome and browser drivers.

**Comment by nturgut** [comment]
I created this issue for writing plist file: https://github.com/flutter/flutter/issues/62054

**Comment by digiter** [comment]
The plist approach doesn't seem to work, @nturgut please take a look at https://chromium-swarm.appspot.com/task?id=4d90f4304f4f1310

**Comment by digiter** [comment]
> From another thread with @nturgut , we can help you run `/usr/bin/safaridriver --enable` command on "flutter-try-mac-X" bots. Does it sound good to you?

I think we can give this approach a try.

**Comment by godofredoc** [comment]
Applying salt states fail with:

```
----------
          ID: /usr/local/bin/validate_json.py
    Function: file.managed
      Result: False
     Comment: Group root is not available
     Started: 16:06:23.764596
    Duration: 2.494 ms
     Changes:   
----------
```

@digiter do you know what that means?

**Comment by digiter** [comment]
I think it's irrelevant to our salt states. Maybe a bug from Fuchsia states.

**Comment by godofredoc** [comment]
I don't think your change was applied, as the machines are still asking for the password. Either the change was not cherry picked or salt if failing to apply the states. Let me validate the change is in salt prod

**Comment by godofredoc** [comment]
Some more info from @nturgut:

```
I completed the test we discussed yesterday.

The bots I used: flutter-try-mac-4 flutter-try-mac-9

Results: 
https://chromium-swarm.appspot.com/task?id=4ddafecb118d7b10
https://chromium-swarm.appspot.com/task?id=4ddafe1d98df6610

The recipe steps I added:
  if api.platform.is_mac:
    api.step('sudo enable driver', ['/usr/bin/sudo', '/usr/bin/safaridriver',
             '--enable'])
    api.step('run safari driver', ['/usr/bin/safaridriver', '--port=4444', '&'])
    api.step('system_profiler', ['system_profiler', 'SPApplicationsDataType',
             '-xml'])

The output of `enable` step is not seen. Yet since the other command failed, I guess Safari is not enabled in these bots.

Please let me know if there is anything else I can help with.
```

**Comment by godofredoc** [comment]
@nturgut these is the sequence of commands that work for me:

```
bash-3.2$ sudo /usr/bin/safaridriver --enable
bash-3.2$ /usr/bin/safaridriver --port=4444 & 
```

Which returns:

```
[1] 5632
bash-3.2$ system_profiler SPApplicationsDataType -xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<array>
        <dict>
                <key>_SPCommandLineArguments</key>
                <array>
                        <string>/usr/sbin/system_profiler</string>
                        <string>-nospawn</string>
                        <string>-xml</string>
                        <string>SPApplicationsDataType</string>
                        <string>-detailLevel</string>
                        <string>full</string>
                </array>
                <key>_SPCompletionInterval</key>
                <real>4.6869269609451294</real>
                <key>_SPResponseTime</key>
                <real>4.7800090312957764</real>
                <key>_dataType</key>
                <string>SPApplicationsDataType</string>
                <key>_detailLevel</key>
                <integer>1</integer>
....
```

**Comment by godofredoc** [comment]
@nturgut please give it a try using 'sudo' without the path prefix and let us know if something else is needed.

**Comment by nturgut** [comment]
@godofredoc @digiter I still get the same error when I write the recipe to use only 'sudo'. When executing the command it automatically ads /usr/bin in front of it.

Results: https://chromium-swarm.appspot.com/task?id=4de2bf11a8444410
https://chromium-swarm.appspot.com/task?id=4de2bf5b904d1410

applied recipe steps:
```
  if api.platform.is_mac:
    api.step('sudo enable driver', ['sudo', '/usr/bin/safaridriver',
             '--enable'])
    api.step('run safari driver', ['/usr/bin/safaridriver', '--port=4444', '&'])
    api.step('system_profiler', ['system_profiler', 'SPApplicationsDataType',
             '-xml'])
```

Can you also test this in a recipe? Please let me know if there is anything I can help.

**Comment by godofredoc** [comment]
I don't think this one should be in the infra ticket queue as it has been open for a long time.

**Comment by digiter** [comment]
Found a related SO, we probably need to create an application https://stackoverflow.com/questions/48262318/unable-to-run-tests-against-safari-11-remotely-ssh-ci

**Comment by godofredoc** [comment]
Had some time to investigate about this. Short answer is  using safari driver over ssh is not supported as it requires an active session.

The workaround @digiter mentioned works sometimes but after killing the app for the first time it stops working for some random time.

@nturgut  I'd like to close this as unfeasible unless web team would like to spend some more time to make the work in a reliable way.

**Comment by nturgut** [comment]
@godofredoc I'm closing this issue for now. I'll look for alternatives to run integration tests for Safari and open another issue.
